- Type

        1. Integer
        2. Float
        3. Complex

- Operation

        1. numbers              — Numeric abstract base classes
        2. math                 — Mathematical functions
        3. cmath                — Mathematical functions for complex numbers
        4. decimal              — Decimal fixed point and floating point arithmetic
        5. fractions            — Rational numbers
        6. random               — Generate pseudo-random numbers
        7. statistics           — Mathematical statistics functions

# <center> Numeric types</center> 

In [5]:
#Input
b2=int(input("Enter int  :  "))  
type(b2)

Enter int  :  25


int

## 1. int( )

In [233]:
#help(int)

In [234]:
b=int()
type(b)

int

In [235]:
# Calling function
val = int(10) # integer value
val2 = int(10.52) # float value
val3 = int('10') # string value
# Displaying result
print("integer values :",val, val2, val3) 

integer values : 10 10 10


In [236]:
b1=10
b1

10

In [237]:
type(b1)

int

In [7]:
y='0xc545d0'
int(y,base=16)

12928464

In [6]:
y='0b10100'
int(y,base=2)

20

## c. float( )

In [ ]:
#help(float)

In [ ]:
c=float()
type(c)

In [ ]:
c1=10.5
print(c1)
c1

In [ ]:
type(c1)

In [ ]:
c2 = float(input("Enter FLOAT  :  "))
c2

## d. complex( )

In [2]:
#help(complex)

In [9]:
d=complex(input())
type(d)

2.0+2.3j


complex

In [ ]:
# Python complex() function example
# Calling function
a = complex(1) # Passing single parameter
b = complex(1,2) # Passing both parameters
# Displaying result
print(a)
print(b) 

In [ ]:
d1=2.0 + 2.3j
print(d1)
d1

In [ ]:
#Complex
x = 3+5j
y = 5j
z = -5j

In [ ]:
type(d1)

In [ ]:
d2 = complex(input("Enter COMPILE  :  "))     #2+2.3j
d2

# <center>Type Casting</center>


1. int()     - Integer
2. float()   - Float Value
3. chr()     - Charecter Value
4. complex

In [10]:
#Integers:
int(1)   # x will be 1

1

In [11]:
int(2.8) # y will be 2

2

In [12]:
int("3") # z will be 3

3

In [14]:
#Floats:
float(1)     # x will be 1.0

1.0

In [16]:
float(2.8)   # y will be 2.8

2.8

In [17]:
float("3")   # z will be 3.0

3.0

In [18]:
float("4.2") # w will be 4.2

4.2

**Note :**
    
 - can't convert complex to int 
 - invalid literal for int() with base 10:

# <center>Built-in Functions</center>


In [ ]:
#help(__builtins__)

## 3. Math fuction

**abs()**

In [ ]:
# help(abs)          Remove Sign

In [64]:
print('Absolute value of -40 is:', abs(-20))

Absolute value of -40 is: 20


In [66]:
print('Absolute value of -40.83 is:', abs(-20.83))

Absolute value of -40.83 is: 20.83


**sum()**

In [ ]:
#help(sum)

In [62]:
sum([1, 2,4 ])

7

In [61]:
sum([1, 2, 4], 10)

17

**round()**

In [ ]:
#help(round)

In [58]:
round(10)  

10

In [59]:
round(10.8)

11

In [60]:
round(6.6)

7

**pow() / Root**

In [ ]:
#help(pow)
# positive x, positive y (x**y)  

In [67]:
pow(4, 2)

16

In [68]:
pow(-4, 2)

16

In [69]:
pow(4, -2)

0.0625

In [70]:
pow(-4, -2)

0.0625

**divmod()**

In [ ]:
#help(divmod)

In [74]:
divmod(11,2)

(5, 1)

In [75]:
divmod(11,2)[0]

5

In [76]:
divmod(11,2)[1]

1

**eval()**

In [ ]:
#help(eval)

In [77]:
x = 8
eval('x + 1')

9

# Other Lib / Operation

In [ ]:
https://mathworld.wolfram.com/topics/CombinatorialGeometry.html
    
https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.jamesbrennan.org%2Falgebra%2Fnumbers%2Freal_number_system.htm&psig=AOvVaw0K2qi1v15kWLI96oGJ4lHI&ust=1596723933259000&source=images&cd=vfe&ved=0CAIQjRxqFwoTCPDSraqihOsCFQAAAAAdAAAAABAO

## 1. numbers

- A complex number is a number that can be expressed in the form a + bi, where a and b are real numbers, and i satisfies the equation i2 = −1. Because no real number satisfies this equation, i is called an imaginary number. For the complex number a + bi, a is called the real part, and b is called the imaginary part.

- Number
    - Complex
        - Real
            - Rational
                 - Integral

In [18]:
import numbers
# help(numbers)

In [19]:
# class Complex(Number) 
x = 3+5j

In [20]:
x.real ,x.imag 

(3.0, 5.0)

In [21]:
# class Integral(Rational)
x=20

In [17]:
x.denominator,x.numerator

(1, 20)

## 2. math

In [42]:
from math import *
import math as  mt

In [45]:
# help(mt)

In [46]:
print("pi   :",pi)       #3.141592653589793
print("e    :",e)        #2.718281828459045
print("inf  :",inf)      #inf
print("nan  :",nan)      #nan
print("tau  :",tau)      #6.283185307179586

pi   : 3.141592653589793
e    : 2.718281828459045
inf  : inf
nan  : nan
tau  : 6.283185307179586


In [47]:
# isfinite(x)  : Return True if x is neither an infinity nor a NaN, and False otherwise.
# isinf(x)     : Return True if x is a positive or negative infinity, and False otherwise.
# isnan(x)     : Return True if x is a NaN (not a number), and False otherwise.

**A. Basic**

- **floor(x):** Return the floor of x as an Integral.
    - This is the largest integer <= x.

In [48]:
floor(23.255)

23

- **fmod(x, y):**Return fmod(x, y), according to platform C.
    - x % y may differ.

In [49]:
fmod(200,13)

5.0

- **exp(x):**Return e raised to the power of x.

In [50]:
exp(2)

7.38905609893065

- **pow(x, y):** Return x**y (x to the power of y).

- **sqrt(x):** Return the square root of x.

In [51]:
sqrt(9)

3.0

- **modf(x, /):** Return the fractional and integer parts of x.
    - Both results carry the sign of x and are floats.

In [52]:
modf(2/3)

(0.6666666666666666, 0.0)

- **factorial(x):**Find x!.

In [53]:
factorial(5)

120

### Function

- **ceil(x, /):**Return the ceiling of x as an Integral.
    - his is the smallest integer >= x.

- **copysign(x, y, /):** Return a float with the magnitude (absolute value) of x but the sign of y.
    - On platforms that support signed zeros, copysign(1.0, -0.0)
    - returns -1.0.

- **lgamma(x):** Natural logarithm of absolute value of Gamma function at x.
    - **erf(x):**Error function at x.

 - **erfc(x, /):** Complementary error function at x.

- **expm1(x, /):Return exp(x)-1.**
    - This function avoids the loss of precision involved in the direct evaluation of exp(x)-1 for small x.

- **fabs(x, /):** Return the absolute value of the float x.

- **frexp(x, /):**Return the mantissa and exponent of x, as pair (m, e).
    - m is a float and e is an int, such that x = m * 2.**e.
    - If x is 0, m and e are both 0.  Else 0.5 <= abs(m) < 1.0.

- **fsum(seq, /):** Return an accurate floating point sum of values in the iterable seq.
    - Assumes IEEE-754 floating point arithmetic.

- **gamma(x, /):** Gamma function at x.

- **gcd(x, y, /):** greatest common divisor of x and y

- **hypot(x, y, /):** Return the Euclidean distance, sqrt(x*x + y*y).

- **isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0):** Determine whether two floating point numbers are close in value.

- **ldexp(x, i, /):** Return x * (2**i).
    - This is essentially the inverse of frexp().
    

- **remainder(x, y, /):**
    - Difference between x and the closest integer multiple of y.
    - Return x - n*y where n*y is the closest integer multiple of y.
    - In the case where x is exactly halfway between two multiples of y, the nearest even value of n is used. The result is always exact.

- **trunc(x, /)**
    - Truncates the Real x to the nearest Integral toward 0.
    - Uses the __trunc__ magic method.

**B.Trignomatric**

> π rad = 180°
>
> 1Rad × 180/π = 57.296Deg
>
> α(degrees) = α(radians) × 180° / π
>
> degrees = radians × 180° / π

|Radians (rad)|Radians (rad)|Degrees (°)|
|:---|:---|:---|
|0 |0 |0°|
|π/6 |0.5235987756 |30°|
|π/4 |0.7853981634 |45°|
|π/3 |1.0471975512 |60°|
|π/2 |1.5707963268 |90°|
|2π/3 |2.0943951024 |120°|
|3π/4 |2.3561944902 |135°|
|5π/6 |2.6179938780 |150°|
|π |3.1415926536 |180°|
|3π/2 |4.7123889804 |270°|
|2π |6.2831853072 |360°|

- **degrees(x):** Convert angle x from radians to degrees.

In [58]:
degrees(1.5707963267948966)

90.0

- **radians(x):** Convert angle x from degrees to radians.
    

In [59]:
radians(90)

1.5707963267948966

- **sin(x) :** Return the sine of x (measured in radians). 

In [60]:
sin(1.5707963267948966)

1.0

In [61]:
sin(radians(90))

1.0

In [62]:
sin(radians(0))

0.0

- **cos(x) :** Return the cosine of x (measured in radians).

In [79]:
cos(0)

1.0

In [83]:
cos(1.5707963267948966)

6.123233995736766e-17

In [78]:
cos(radians(90)

6.123233995736766e-17

In [74]:
cos(0)

1.0

In [75]:
cos(1)

0.5403023058681398

- **tan(x, /):** Return the tangent of x (measured in radians).

- **asin(x, /):** Return the arc sine (measured in radians) of x.  

- **acos(x, /):** Return the arc cosine (measured in radians) of x.    

- **atan(x, /):** Return the arc tangent (measured in radians) of x.

- **atan2(y, x, /):** Return the arc tangent (measured in radians) of y/x.
    - Unlike atan(y/x), the signs of both x and y are considered.
    

**hyperbolic**

- **sinh(x, /):** Return the hyperbolic sine of x.

- **cosh(x, /):** Return the hyperbolic cosine of x.

- **tanh(x, /):** Return the hyperbolic tangent of x.

- **asinh(x, /):** Return the inverse hyperbolic sine of x.

- **acosh(x, /):** Return the inverse hyperbolic cosine of x.

- **atanh(x, /):** Return the inverse hyperbolic tangent of x.
    

### log(...)

- **log(x, [base=math.e]):** Return the logarithm of x to the given base.
    - If the base not specified, returns the natural logarithm (base e) of x.

In [84]:
log(10)

2.302585092994046

- **log10(x)**: Return the base 10 logarithm of x.
    

In [86]:
log10(10)

1.0

- **log1p(x)**: Return the natural logarithm of 1+x (base e).
    - The result is computed in a way which is accurate for x near zero.

In [ ]:
log1p(10)

- **log2(x)**: Return the base 2 logarithm of x.
    

In [87]:
log2(2)

1.0

## [cmath](https://docs.python.org/3.7/library/cmath.html)

cmath.phase(x)|Return the phase of x (also known as the argument of x), as a float. phase(x) is equivalent to math.atan2(x.imag, x.real). The result lies in the range [-π, π], and the branch cut for this operation lies along the negative real axis, continuous from above. On systems with support for signed zeros (which includes most systems in current use), this means that the sign of the result is the same as the sign of x.imag, even when x.imag is zero:

Note The modulus (absolute value) of a complex number x can be computed using the built-in abs() function. There is no separate cmath module function for this operation.
cmath.polar(x)|Return the representation of x in polar coordinates. Returns a pair (r, phi) where r is the modulus of x and phi is the phase of x. polar(x) is equivalent to (abs(x), phase(x)).
|cmath.rect(r, phi)|Return the complex number x with polar coordinates r and phi. Equivalent to r * (math.cos(phi) + math.sin(phi)*1j).

**2. Power and logarithmic functions**\

|Function| Description|
|:-----|:------|
|cmath.exp(x)|Return e raised to the power x, where e is the base of natural logarithms.
|cmath.log(x[, base])|Returns the logarithm of x to the given base. If the base is not specified, returns the natural logarithm of x. There is one branch cut, from 0 along the negative real axis to -∞, continuous from above.
|cmath.log10(x)|Return the base-10 logarithm of x. This has the same branch cut as log().
|cmath.sqrt(x)|Return the square root of x. This has the same branch cut as log().

**3. Trigonometric functions**

|Function| Description|
|:-----|:------|
|cmath.acos(x)|Return the arc cosine of x. There are two branch cuts: One extends right from 1 along the real axis to ∞, |continuous from below. The other extends left from -1 along the real axis to -∞, continuous from above.
|cmath.asin(x)|Return the arc sine of x. This has the same branch cuts as acos().
|cmath.atan(x)|Return the arc tangent of x. There are two branch cuts: One extends from 1j along the imaginary axis to ∞j, continuous from the right. The other extends from -1j along the imaginary axis to -∞j, continuous from the left.
|cmath.cos(x)|Return the cosine of x.
|cmath.sin(x)|Return the sine of x.
|cmath.tan(x)|Return the tangent of x.


**4. Hyperbolic functions**

|Function| Description|
|:-----|:------|
|cmath.acosh(x)|Return the inverse hyperbolic cosine of x. There is one branch cut, extending left from 1 along the real axis to -∞, continuous from above.
|cmath.asinh(x)|Return the inverse hyperbolic sine of x. There are two branch cuts: One extends from 1j along the imaginary axis to ∞j, continuous from the right. The other extends from -1j along the imaginary axis to -∞j, continuous from the left.
|cmath.atanh(x)|Return the inverse hyperbolic tangent of x. There are two branch cuts: One extends from 1 along the real axis to ∞, continuous from below. The other extends from -1 along the real axis to -∞, continuous from above.
|cmath.cosh(x)|Return the hyperbolic cosine of x.
|cmath.sinh(x)|Return the hyperbolic sine of x.
|cmath.tanh(x)|Return the hyperbolic tangent of x.

**5. Classification functions**

|Function| Description|
|:-----|:------|
|cmath.isfinite(x)|Return True if both the real and imaginary parts of x are finite, and False otherwise.
|cmath.isinf(x)|Return True if either the real or the imaginary part of x is an infinity, and False otherwise.
|cmath.isnan(x)|Return True if either the real or the imaginary part of x is a NaN, and False otherwise.
|cmath.isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)|Return True if the values a and b are close to each other and False otherwise.


**6. Constants**

|Function| Description|
|:-----|:------|
|cmath.pi|The mathematical constant π, as a float.
|cmath.e|The mathematical constant e, as a float.
|cmath.tau|The mathematical constant τ, as a float.
|cmath.inf|Floating-point positive infinity. Equivalent to float('inf').
|cmath.infj|Complex number with zero real part and positive infinity imaginary part. Equivalent to complex(0.0, float('inf')).
|cmath.nan|A floating-point “not a number” (NaN) value. Equivalent to float('nan').
|cmath.nanj|Complex number with zero real part and NaN imaginary part. Equivalent to complex(0.0, float('nan')).

## 4. decimal

In [98]:
from decimal import *

In [99]:
Decimal('10')

Decimal('10')

In [100]:
Decimal('3.14')

Decimal('3.14')

In [102]:
Decimal(22/7)

Decimal('3.142857142857142793701541449991054832935333251953125')

In [103]:
Decimal(3.14)

Decimal('3.140000000000000124344978758017532527446746826171875')

## 5. fractions

In [89]:
from fractions import Fraction

In [90]:
Fraction(16, -10)

Fraction(-8, 5)

In [91]:
Fraction(123)

Fraction(123, 1)

In [92]:
Fraction()

Fraction(0, 1)

In [93]:
Fraction('3/7')

Fraction(3, 7)

In [94]:
Fraction('1.414213')

Fraction(1414213, 1000000)

In [95]:
Fraction('7e-6')

Fraction(7, 1000000)

## 6. random

|Method|Descriptio|
|:---|:---|
|seed()|Initialize the random number generator
|getstate()|Returns the current internal state of the random number generator
|setstate()|Restores the internal state of the random number generator
|getrandbits()|Returns a number representing the random bits
|randrange()|Returns a random number between the given range
|randint()|Returns a random number between the given range
|choice()|Returns a random element from the given sequence
|choices()|Returns a list with a random selection from the given sequence
|shuffle()|Takes a sequence and returns the sequence in a random order
|sample()|Returns a given sample of a sequence
|random()|Returns a random float number between 0 and 1
|uniform()|Returns a random float number between two given parameters
|triangular()|Returns a random float number between two given parameters, you can also set a mode parameter to specify the midpoint between the two other parameters
|betavariate()|Returns a random float number between 0 and 1 based on the Beta distribution (used in statistics)
|expovariate()|Returns a random float number based on the Exponential distribution (used in statistics)
|gammavariate()|Returns a random float number based on the Gamma distribution (used in statistics)
|gauss()|Returns a random float number based on the Gaussian distribution (used in probability theories)
|lognormvariate()|Returns a random float number based on a log-normal distribution (used in probability theories)
|normalvariate()|Returns a random float number based on the normal distribution (used in probability theories)
|vonmisesvariate()|Returns a random float number based on the von Mises distribution (used in directional statistics)
|paretovariate()|Returns a random float number based on the Pareto distribution (used in probability theories)
|weibullvariate()|Returns a random float number based on the Weibull distribution (used in statistics)

In [104]:
import random

In [124]:
sample_list=[1, 2, 3, 4, 5, 6]
random.shuffle(sample_list)
sample_list

[6, 4, 3, 1, 2, 5]

In [118]:
random.choice([1, 4, 8, 10, 3])

10

In [116]:
random.randrange(20, 50, 3)

20

In [120]:
random.seed(5)